backend/push.js

```diff
@@ -1,53 +1,332 @@
+    // it likely means that the subscription no longer exists.
+    if ([401, 404, 410].includes(req.status)) {
+      console.warn(
+        new Date().toISOString(),
+        'Removing subscription',
+        subscription.id
+      )
+      deleteClient(subscription.id)
+      return
+    }
+    if (req.status === 413) {
+      throw new Error('Payload too large')
+    }
+  }
+}
 
 export const pushServerActionhandlers: any = {
   [PUSH_SERVER_ACTION_TYPE.SEND_PUBLIC_KEY] () {
     const socket = this
-    socket.send(createMessage(REQUEST_TYPE.PUSH_ACTION, publicKey))
+    socket.send(createMessage(REQUEST_TYPE.PUSH_ACTION, { type: PUSH_SERVER_ACTION_TYPE.SEND_PUBLIC_KEY, data: getVapidPublicKey() }))
   },
-  [PUSH_SERVER_ACTION_TYPE.STORE_SUBSCRIPTION] (payload) {
+  async [PUSH_SERVER_ACTION_TYPE.STORE_SUBSCRIPTION] (payload) {
     const socket = this
-    const subscription = JSON.parse(payload)
+    const { server } = socket
+    const subscription = payload
+    const subscriptionId = await getSubscriptionId(subscription)
 
-    // Reference: is it safe to use 'endpoint' as a unique identifier of a push subscription
-    // (https://stackoverflow.com/questions/63767889/is-it-safe-to-use-the-p256dh-or-endpoint-keys-values-of-the-push-notificatio)
-    socket.server.pushSubscriptions[subscription.endpoint] = subscription
+    if (!server.pushSubscriptions[subscriptionId]) {
+      // If this is a new subscription, we call `subscriptionInfoWrapper` and
+      // store it in memory.
+      server.pushSubscriptions[subscriptionId] = subscriptionInfoWrapper(subscriptionId, subscription)
+      addSubscriptionToIndex(subscriptionId).then(() => {
+        return sbp('chelonia/db/set', `_private_webpush_${subscriptionId}`, JSON.stringify({ subscription: subscription, channelIDs: [] }))
+      }).catch((e) => console.error(e, 'Error saving subscription', subscriptionId))
+      // Send an initial push notification to verify that the endpoint works
+      // This is mostly for testing to be able to auto-remove invalid or expired
+      // endpoints. This doesn't need more error handling than any other failed
+      // call to `postEvent`.
+      postEvent(server.pushSubscriptions[subscriptionId], JSON.stringify({ type: 'initial' })).catch(e => console.warn(e, 'Error sending initial push notification'))
+    } else {
+      // Otherwise, if this is an _existing_ push subscription, we don't need
+      // to all `subscriptionInfoWrapper` but we need to stop sending messages
+      // over the push subscription (since we now have a WS to use, the one
+      // over which the message came).
+      if (server.pushSubscriptions[subscriptionId].sockets.size === 0) {
+        server.pushSubscriptions[subscriptionId].subscriptions.forEach((channelID) => {
+          if (!server.subscribersByChannelID[channelID]) return
+          server.subscribersByChannelID[channelID].delete(server.pushSubscriptions[subscriptionId])
+        })
+      }
```

Comment:
A few issues here:

1. I'm pretty sure `sockets.size` will not be `0` here. It will be at least `1`, so although the comment says "we need to stop sending messages over the push subscription", this doesn't seem like it's actually done.
2. Each web socket is supposed to only have a single push subscription associated with it, right? So then why do we even attempt to keep track of these `.sockets`? Perhaps we should just remove this `.sockets` property to simplify the code.

Reply:
This particular code seems to work, but:

1. Why would you think `sockets.size` will be at least 1? I think it's perfectly possible for it to be 1. This would happen if the previous WS was closed. The reason to check if the value is 1 or greater is that this only needs to be done when `size` is 0, which means that we're sending messages over push.
2. `Each web socket is supposed to only have a single push subscription associated with it, right?` Yes, but the assumption is that the inverse doesn't always hold, i.e., each push subscription could have multiple web sockets associated with that. I can't prove that this will not happen. Some reasons could include several service workers running at once, the WS being 'closed' from the client's POV but not from the server's POV, etc.